source/common/http/http2/codec_impl.cc

```diff
@@ -502,19 +505,41 @@ void ConnectionImpl::StreamImpl::decodeData() {
     return;
   }
 
-  // Any buffered data will be consumed.
+  // Some buffered body will be consumed. If there remains buffered body after
+  // this call, set this to true.
   stream_manager_.body_buffered_ = false;
 
+  bool already_drained_data = false;
   // It's possible that we are waiting to send a deferred reset, so only raise data if local
   // is not complete.
   if (!deferred_reset_) {
-    decoder().decodeData(*pending_recv_data_, sendEndStream());
+    const bool decode_data_in_chunk =
+        defer_processing_backedup_streams_ && stream_manager_.decodeAsChunks() &&
+        pending_recv_data_->length() > stream_manager_.defer_processing_segment_size_;
+    if (decode_data_in_chunk) {
+      Buffer::OwnedImpl chunk_buffer;
+      // TODO(kbaichoo): Consider implementing an approximate move for chunking.
+      chunk_buffer.move(*pending_recv_data_, stream_manager_.defer_processing_segment_size_);
+
+      // With the current implementation this should always be true,
+      // though this can change with approximation.
+      stream_manager_.body_buffered_ = true;
+
+      decoder().decodeData(chunk_buffer, sendEndStream());
+      already_drained_data = true;
+
+      if (!buffersOverrun()) {
+        process_buffered_data_callback_->scheduleCallbackNextIteration();
+      }
+    } else {
+      // Send the entire buffer through.
+      decoder().decodeData(*pending_recv_data_, sendEndStream());
+    }
   }
 
-  // TODO(kbaichoo): If dumping buffered data, we should do so in default read
-  // size chunks rather than dumping the entire buffer, which can have fairness
-  // issues.
-  pending_recv_data_->drain(pending_recv_data_->length());
+  if (!already_drained_data) {
+    pending_recv_data_->drain(pending_recv_data_->length());
```

Comment:
hm, intuitively I'd think we could put this with the decodeData branch in 536 but after more than 10s thought that would mean we don't drain if there's a deferred_reset_
should we drain in that case?